aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java

```diff
@@ -88,23 +96,69 @@ public void seek(long newPos) {
 
   @Override
   public int read() throws IOException {
-    Preconditions.checkState(!closed, "Cannot read: already closed");
-    positionStream();
+    int[] byteRef = new int[1];
```

Comment:
This is less than ideal but Tasks is a little limited in this area because it's really based on threaded execution and we're reusing it here for retry.  It might be worth exploring whether we can tweak Tasks to support this usecase:

Ideally we'd have something like:

```java
int read = Tasks.single()
    .exponentialBackoff(...)
    ...
    run(<funciton>);
```

Reply:
We could also consider bringing in [Failsafe](https://failsafe.dev/), which is more closely aligned with what we want for these cases.  Though it is a new dependency (but zero deps and apache licensed).